src/corecel/cont/Bitset.hh

```diff
@@ -0,0 +1,553 @@
+//! Bitwise AND with another bitset
+template<size_type N>
+CELER_CONSTEXPR_FUNCTION Bitset<N>&
+Bitset<N>::operator&=(Bitset const& other) noexcept
+{
+    for (size_type i = 0; i < num_words; ++i)
+    {
+        words_[i] &= other.words_[i];
+    }
+    return *this;
+}
+
+//---------------------------------------------------------------------------//
+//! Bitwise OR with another bitset
+template<size_type N>
+CELER_CONSTEXPR_FUNCTION Bitset<N>&
+Bitset<N>::operator|=(Bitset const& other) noexcept
+{
+    for (size_type i = 0; i < num_words; ++i)
+    {
+        words_[i] |= other.words_[i];
+    }
+    return *this;
+}
+
+//---------------------------------------------------------------------------//
+//! Bitwise XOR with another bitset
+template<size_type N>
+CELER_CONSTEXPR_FUNCTION Bitset<N>&
+Bitset<N>::operator^=(Bitset const& other) noexcept
+{
+    for (size_type i = 0; i < num_words; ++i)
+    {
+        words_[i] ^= other.words_[i];
+    }
+    return *this;
+}
+
+//---------------------------------------------------------------------------//
+//! Return a copy with all bits flipped (bianry NOT)
+template<size_type N>
+CELER_CONSTEXPR_FUNCTION Bitset<N> Bitset<N>::operator~() const noexcept
+{
+    return Bitset{*this}.flip();
+}
+
+//---------------------------------------------------------------------------//
+//! Set all bits
+template<size_type N>
+CELER_CONSTEXPR_FUNCTION Bitset<N>& Bitset<N>::set() noexcept
+{
+    for (size_type i = 0; i < num_words; ++i)
+    {
+        words_[i] = ~static_cast<word_type>(0);
+    }
+    // sanitize the last word
+    detail::Sanitize<N % bits_per_word>::sanitize(this->last_word());
+    return *this;
+}
+
+//---------------------------------------------------------------------------//
+//! Set the bit at position pos
+template<size_type N>
+CELER_CONSTEXPR_FUNCTION Bitset<N>&
+Bitset<N>::set(size_type pos, bool value) noexcept
+{
+    (*this)[pos] = value;
+    return *this;
+}
+
+//---------------------------------------------------------------------------//
+//! Reset all bits
+template<size_type N>
+CELER_CONSTEXPR_FUNCTION Bitset<N>& Bitset<N>::reset() noexcept
+{
+    for (size_type i = 0; i < num_words; ++i)
+    {
+        words_[i] = static_cast<word_type>(0);
+    }
+
+    return *this;
+}
+
+//---------------------------------------------------------------------------//
+//! Reset the bit at position pos
+template<size_type N>
+CELER_CONSTEXPR_FUNCTION Bitset<N>& Bitset<N>::reset(size_type pos) noexcept
+{
+    this->get_word(pos) &= ~Bitset::mask(pos);
+    return *this;
+}
+
+//---------------------------------------------------------------------------//
+//! Flip all bits
+template<size_type N>
+CELER_CONSTEXPR_FUNCTION Bitset<N>& Bitset<N>::flip() noexcept
+{
+    for (size_type i = 0; i < num_words; ++i)
+    {
+        words_[i] = ~words_[i];
+    }
+    // sanitize the last word
+    detail::Sanitize<N % bits_per_word>::sanitize(last_word());
```

Comment:
Now that we don't support N=0 I think it would be cleaner to make this a member function. (Also note that "sanitize" means "zero unused bits")

Reply:
We still the no-op specialization when `N % bits_per_word=0`, and we can't specialize a member of an unspecialized template, so I think it still needs to be a separate function